docs: ADR: Telemetry Flow Healthiness Status API (#802)
kyma-project/telemetry-manager: docs/contributor/arch/008-telemetry-flow-healthiness-status-api.md

# 8. Telemetry Flow Healthiness Status API

Date: 2024-15-02

## Status

Proposed

## Context

As a follow-up of [ADR 003: Integrate Prometheus With Telemetry Manager Using Alerting](003-integrate-prometheus-with-telemetry-manager-using-alerting.md),
let's actually define the Telemetry Healthiness Status API for Trace and Metric Pipelines.

> In general, our aim is to highlight significant events in the telemetry flow as status conditions. In the event of issues, it should be apparent to the customer where the problem lies.

Reviewer: ```suggestion
In general, our aim is to highlight significant events in the telemetry flow as status conditions. When issues occur, it should be apparent to the customer where the problem lies.
```

> Additionally, we should offer runbooks to assist them in resolving these problems. The diagram below visually represents these key events:

Reviewer: ```suggestion
Additionally, we should offer runbooks to help them in resolving these problems. The following diagram shows these key events:
```

![OTel Collector Data Flow](../assets/otel-collector-data-flow.svg "OTel Collector Data Flow")

### Throttling

* Currently, can be only triggered by Memory Limiter (increase in `otelcol_receiver_refused_metric_points` and `otelcol_processor_refused_metric_points`).
There is a community discussion about incorporating a rate-limiting mechanism directly into the OTLP Receiver: https://github.com/open-telemetry/opentelemetry-collector/issues/6725.
* Often triggered by memory pressure from the exporter queue, although during extreme traffic spikes, memory may exceed the soft limit even when the queue is empty.
* gRPC status code Unavailable returned to the client.
* Scaling may help if the backend is healthy, but could worsen things if not.

### Filter and Transform Processors Refusing Data

* Most failures are checked at startup (OTTL syntax), causing collector crashes. Errors may occur with specific functions like ParseJSON().
* When `error_mode == propagate`, the processor refuses data, and a signal goes back to the receiver for client error notification.

> * Not relevant for us so far since we neither use `error_mode == propagate` nor use OTTL functions that can cause errors.

Reviewer: ```suggestion
* Not relevant for us so far, because we neither use `error_mode == propagate` nor use OTTL functions that can cause errors.
```

### Exporter Queueing

* All batches are enqueued first.

> * Queue fills up when consumers are slower than producers. It can be caused by backend issues or a mismatch between the ingestion and export rate (e.g. backend is slow).

Reviewer: ```suggestion
* Queue fills up when consumers are slower than producers. It can be caused by backend issues or a mismatch between the ingestion and export rate (for example, backend is slow).
```

* If the exporter queue is full, data is dropped (`otelcol_exporter_enqueue_failed_metric_points` goes up).

> * Watch for high queue size: `otelcol_exporter_queue_size / otelcol_exporter_queue_capacity > THRESHOLD`.

Reviewer: who watches? the user?

Reviewer 2: This is an expression that would be used to determining high queue size. Here is just an info how to determine high queue size.

### Exporter Retries

> * For non-retryable errors, data is dropped. See more about retryable and non-retryable errors.

Reviewer: should this have a link?

Reviewer: The sentence was dropped.

* Retryable errors trigger retries until success or retry limit, then data is dropped.

> * `otelcol_exporter_send_failed_metric_points` is increased in both scenarios if data is dropped.

Reviewer: ```suggestion
* For retryable and non-retryable errors: If data is dropped, `otelcol_exporter_send_failed_metric_points` is increased.
```

* If data is sent successfully, `otelcol_exporter_sent_metric_points` is increased.

## Decision

We are choosing between two alternatives:

* Using multiple condition types to represent various telemetry flow events (Throttling, Data Loss, High Buffer Utilization, etc.).

> * Using a single condition type (TelemetryFlowHealthy) with a reason field to denote diverse telemetry flow events.

Reviewer: ```suggestion
* Using a single condition type (`TelemetryFlowHealthy`) with a reason field to denote diverse telemetry flow events.
```

> After careful consideration, we have opted for the single condition type TelemetryFlowHealthy as it minimizes cognitive load for the end user.

Reviewer: ```suggestion
After careful consideration, we have opted for the single condition type `TelemetryFlowHealthy`, because it is easier to understand for the end user.
```

> Ultimately, the user's primary concern is understanding whether the telemetry flow is functioning correctly. In case of any issues, the user has the following actionable steps:

Reviewer: ```suggestion
Ultimately, the user's primary concern is understanding whether the telemetry flow is functioning correctly. In case of any issues, the user can do the following:
```

* Troubleshoot the backend.
* Investigate backend connectivity.
* Reduce ingestion.
* Manually scale out the gateway (as long as no autoscaling capability is provided).

> In this scenario, we can assign the reason field a value that holds utmost significance for the user:

Reviewer: ```suggestion
With a single condition type, we can assign the **reason** field a value that is most relevant for the user:
```

> With a single condition type, we can assign the **reason** field a value that is most relevant for the user:

Reviewer: ```suggestion
With a single condition type, the **reason** field will show a value that is most relevant for the user. We suggest the following values, ordered from most to least critical:
```

```

> FullDataLoss > PartialDataLoss > HighBufferUtilization > GatewayThrottling > Healthy

Reviewer 2: ```suggestion
CompleteDataLoss > PartialDataLoss > HighBufferUtilization > GatewayThrottling > Healthy
```

Reviewer: Just guessing that "HighBufferUtilization" refers to the queue size as indicated in line 33ff.

```suggestion
AllTelemetryDataDropped > SomeTelemetryDataDropped > QueueFillingUp > GatewayThrottling > Healthy
```

Btw, why are these items formatted as code, and separated by `>` ? What are we expressing with this? The sentence before indicates that it could be ordered by relevance to the user, but it rather looks like this is ordered in grades of severity?

Reviewer 2: `>` means severity `AllTelemetryDataDropped` is `critical`

Reviewer: Then let's state that explicitly. See https://github.com/kyma-project/telemetry-manager/pull/802/files#r1504022811

Reviewer 2: I would skip `QueueFillingUp` since we dont want to expose internal functioning for the component. Would stick to `GatewayThrottling`

Reviewer: Alright!

```

The reasons are based on the following alert rules (an example for Metric Pipelines; the actual PromQL expressions must be defined later):
| Alert Rule | Expression |
| --- | --- |
| GatewayExporterSentMetrics    | `sum(rate(otelcol_exporter_sent_metric_points{...}[5m])) > 0`           |
| GatewayExporterDroppedMetrics  | `sum(rate(otelcol_exporter_send_failed_metric_points{...}[5m])) > 0`    |
| GatewayReceiverRefusedMetrics  | `sum(rate(otelcol_receiver_refused_metric_points{...}[5m])) > 0`        |
| GatewayExporterEnqueueFailed   | `sum(rate(otelcol_exporter_enqueue_failed_metric_points{...}[5m])) > 0` |
| GatewayExporterQueueAlmostFull | `otelcol_exporter_queue_size / otelcol_exporter_queue_capacity > 0.8`   |                               |

> We could then map the alert rules to the reasons as follows:

Reviewer: ```suggestion
Then, we map the alert rules to the reasons as follows:
```

| Reason | Alert Rules |
| --- | --- |

> | FullDataLoss           | **not** GatewayExporterSentMetrics **and** (GatewayExporterDroppedMetrics **or** GatewayExporterEnqueueFailed) |

Reviewer 2: ```suggestion
| CompleteDataLoss           | **not** GatewayExporterSentMetrics **and** (GatewayExporterDroppedMetrics **or** GatewayExporterEnqueueFailed) |
```
`FullDataLoss` sounds less professional

Reviewer: If you like my suggestions in line 64, let's repeat them here.

| PartialDataLoss        | GatewayExporterSentMetrics **and** (GatewayExporterDroppedMetrics **or** GatewayExporterEnqueueFailed)       |

> | HighBufferUtilization  | GatewayExporterQueueAlmostFull                                                                           |

Reviewer: I'm not really sure what "HighBufferUtilization" means. In line 64, I'm suggesting "QueueFillingUp" instead. Or maybe "BufferFillingUp"?

| GatewayThrottling      | GatewayReceiverRefusedMetrics                                                                            |

> | Healthy                | **not** (GatewayExporterDroppedMetrics **or** GatewayExporterEnqueueFailed **or** HighBufferUtilization **or** GatewayReceiverRefusedMetrics) |

Reviewer 2: We would eventually harmonize names for all signals so `GatewayThrottling` should be fine.

### Gateway/Agent Communication

> The mentioned events are about the Gateway Collector. Some issues might happen with the Agent, like:

Reviewer: ```suggestion
The events mentioned above are about the Gateway Collector. Some other issues might happen with the Agent, like:
```

> >BufferFillingUp should not result in negative condition status. This reason would be aggregated as warning in telemetry module status

Reviewer: ```suggestion
> **NOTE:** `BufferFillingUp` should not result in negative condition status. This reason would be aggregated as warning in the Telemetry module status.
```

* Gateway can't be reached (network problem).
* Gateway won't accept data (gateway throttling).

> In both cases, the Agent might drop data after too many retries. But since users usually can't do much about these issues, we won't include them in the TelemetryFlowHealthy condition.

Reviewer: ```suggestion
In both cases, the Agent might drop data after too many retries. But because users usually can't do much about these issues, we won't include them in the `TelemetryFlowHealthy` condition.
```

> In both cases, the Agent might drop data after too many retries. But because users usually can't do much about these issues, we won't include them in the `TelemetryFlowHealthy` condition.

Reviewer: Does this mean the user will experience issues in the Telemetry flow, even though `TelemetryFlowHealthy` shows it's all fine?

Reviewer: This feels kind of misleading :D

If there's throttling, scaling horizontally might help, but it's not much different from dealing with a high OTLP ingestion rate.

> In the future, we might think about adding a TelemetryCollectionHealthy condition for the Agent. This would help track issues like scraping problems by Prometheus Receiver.

Reviewer: ```suggestion
In the future, we might think about adding a `TelemetryCollectionHealthy` condition for the Agent. This would help track issues like scraping problems by Prometheus Receiver.
```

## Consequences

> The suggested API addresses possible obstacles in the telemetry flow. We can enhance the information in the message field and include a troubleshooting guide.

Reviewer: ```suggestion
The Telemetry Health Status API addresses possible obstacles in the telemetry flow. We can enhance the information in the message field and include a troubleshooting guide.
```

> Looking ahead, as the OTel Collector evolves, we may introduce additional reasons, as well as expose simplified custom metrics derived from OTel Collector metrics to the user.

Reviewer: ```suggestion
Looking ahead, as the OTel Collector evolves, we may introduce additional values for **reason**, as well as expose simplified custom metrics derived from OTel Collector metrics to the user.
```
